ARP protection

mbssrc opened the pull request:
<!--
    Copyright 2022-2025 TII (SSRC) and the Ghaf contributors
    SPDX-License-Identifier: CC-BY-SA-4.0
-->

## Description of Changes

<!--
Summary of the proposed changes in the PR description in your own words. For dependency updates, please link to the changelog.
-->

Enable static ARP in the internal network, to prevent ARP spoofing and
subsequent MITM attacks.

- additional firewall (ebtables) rules on internal bridge, filtering out
  arp traffic; and required kernel modules
- sysctl settings to configure arp behaviour on bridge and vm internal
  interfaces
- splitting host networking configuration, allowing to later remove host
  networking without removing the internal bridge
- adding static entries for all known hosts
- adding flag for external host networking, allowing to not activate
  interfaces on the host by default and disable internal host nat


### Type of Change
- [ ] New Feature
- [ ] Bug Fix
- [x] Improvement / Refactor

## Related Issues / Tickets

<!--
Link to GitHub issues or JIRA tickets (if any) that this PR addresses or is related to
-->

## Checklist

<!--
Please check [X] for all items that apply. Leave [ ] if an item does not apply, but you have considered it.
Note that none of these are strict requirements — they are intended to inform reviewers.
Completing this checklist shows that you value and respect their time and effort.
-->

- [x] Clear summary in PR description
- [x] Detailed and meaningful commit message(s)
- [x] Commits are logically organized and squashed if appropriate
- [x] [Contribution guidelines](https://github.com/tiiuae/ghaf/blob/main/CONTRIBUTING.md) followed
- [ ] Ghaf documentation updated with the commit - https://tiiuae.github.io/ghaf/
- [ ] Author has run `make-checks` and it passes
- [ ] All automatic GitHub Action checks pass - see [actions](https://github.com/tiiuae/ghaf/actions)
- [x] Author has added reviewers and removed PR draft status

<!-- Additional description of omitted [ ] items if not obvious. -->

## Testing Instructions

### Applicable Targets
- [x] Orin AGX `aarch64`
- [x] Orin NX `aarch64`
- [x] Lenovo X1 `x86_64`
- [x] Dell Latitude `x86_64`

### Installation Method
- [ ] Requires full re-installation
- [x] Can be updated with `nixos-rebuild ... switch`
- [ ] Other: 

### Test Steps To Verify:
<!--
Provide clear, simple step-by-step instructions to verify the functionality.
Please do not assume that readers know everything you currently know.
-->
1. Confirm ARP spoofing works on current main
2. (optional) run `sudo arp net-vm` in chrome-vm before and during spoofing to see MAC address change
3. Apply patch and confirm ARP spoofing does not work (broadcast traffic will still be visible)
4. (optional) run `sudo arp net-vm` in chrome-vm before and during spoofing to see no MAC address change
5. Check for any network performance impact

Instructions:
```
ssh ghaf@zathura-vm
sudo su
nix-shell -p bettercap
arp -s 192.168.100.1 02:ad:00:00:00:01
iptables -P FORWARD ACCEPT

bettercap
> net.probe on
> set arp.spoof.targets 192.168.100.102
> arp.spoof on
> net.sniff on
```

Then open chrome browser on desktop, go to a website and observe traffic in zathura-vm.

enesoztrk commented:
I was considering enabling libvirt's built-in filtering chains, which would prevent spoofing activities between VMs, if integration with Nix is possible

- no-arp-spoofing
- no-ip-spoofing
- no-ipv6-spoofing
- clean-traffic
Reference: https://libvirt.org/formatnwfilter.html#filtering-chains

What do you think?

leivos-unikie commented:
Tested on Lenovo-X1

- instructed arp spoofing test: 
succeeds with ghaf mainline / fails with arp-protect branch

- ci-test-automation ipspoofing test:
succeeds with ghaf mainline / fails with arp-protect branch

- Checked that all vms have internet connection
- Run ci-test-automation network test (iperf) few times, didn't notice clear change compared to mainline
- Desktop and apps seem normal

mbssrc commented on modules/microvm/common/vm-networking.nix:
```diff
@@ -49,11 +52,29 @@ in
         enable = cfg.isGateway;
         internalInterfaces = [ hosts.${cfg.vmName}.interfaceName ];
       };
-      firewall.enable = mkDefault false;
```
Can re-enable this when re-basing onto your patch, but we enable it anyways in the ghaf firewall (which is true by default), so didn't fully understand why to disable it here

enesoztrk replied:
If we want to disable/enable the firewall, it would be good to switch it using `ghaf.firewall.enable` rather `networking.firewall.enable`. Even if we want to disable `ghaf.firewall`, the firewall in ghaf will not be disabled (which is enabled by default)

mbssrc replied:
added back in

leivos-unikie commented:
> > Tested arp spoofing on Orin AGX as below
> > ...
> 
> The host network runs on the bridge, so it can by default see all the traffic - no need for arp spoofing. If you run `net.probe on` and `net.sniff on` you will see the traffic between the VMs (like your curl) already without spoofing.
> 
> What you could do to test the setup on orin:
> 
>     1. Watch the arp entries in admin-vm:
> 
> 
> ```
> watch -n 1 arp
> ```
> 
>     2. Spoof from ghaf-host:
> 
> 
> ```
> (sudo su)
> nix-shell -p bettercap
> arp -s 192.168.100.1 02:ad:00:00:00:01
> iptables -P FORWARD ACCEPT
> bettercap
> # set the target as the admin-vm, and start spoofing
> > set arp.spoof.targets 192.168.100.3 
> > arp.spoof on
> ```
> 
> In main, you should see in the `admin-vm` arp table that the `net-vm` MAC changes to the `ghaf-host` MAC, with the patch nothing should happen. You can also toggle `arp.spoof on` and `arp.spoof off` on the fly to see the difference.
> 
> If you want to dig more into it, you can include the tcpdump package in the build (on-the-fly install will probably not work in admin-vm) and monitor arp traffic on `ethint0` (target VM) and `virbr0` (host bridge). Also, you could try to spoof the host from admin-vm, but again need to include the bettercap package in the build.

I tested again on Orin AGX. 

Toggling net-vm MAC address shown on ghaf-host works in main but nothing happens with your branch as you explained. 

For some reason I was not able to see any http request traffic to ghaf-host with (modified) main image when running bettercap on admin-vm. Maybe if I had set up another vm for running http request it would have worked.

Anyways, the arp protection works as you described both on x86 and ARM device.

mbssrc commented:
Changes:
- rebased with latest firewall changes
- removed additional kernel modules; changed the logic to filter arp traffic on tap-* interfaces. This works on all targets without additional changes. 

useful command on the host to verify the effectiveness: `ebtables -L --Lc` showing the package counter for the rules, can be triggered via arping or actual spoofing. tcpdump shows traffic only on originating interface, both internal bridge and other VM interfaces do not show traffic (`tcpdump -i (ethint0/virbr0/tap-<vm>) -n 'arp'`)

leivos-unikie commented:
Test again also on Orin AGX after the latest changes
- arp spoofing does not succeed (ran bettercap in admin-vm)
- Running `ebtables -L --Lc` in ghaf-host shows the effect of spoofing at firewall
- Updating system time form internet works in ghaf-host and in VMs (NTP service)
- systemctl status does not reveal any services failing in VMs
